draft-ietf-httpbis-message-signatures.md

```diff
@@ -1283,22 +1291,161 @@ The table below contains the initial contents of the HTTP Signature Specialty Co
+
+## Key and Algorithm Specification Downgrades {#security-keydowngrade}
+
+Applications of this specification need to protect against key specification downgrade attacks. For example, the same RSA key can be used for both RSA-PSS and RSA v1.5 signatures. If an application expects a key to only be used with RSA-PSS, it needs to reject signatures for that key using the weaker RSA 1.5 specification.
+
+Another example of a downgrade attack occurs when an asymmetric algorithm is expected, such as RSA-PSS, but an attacker substitutes a signature using symmetric algorithm, such as HMAC. A naive verifier implementation could use the value of the public RSA key as the input to the HMAC verification function. Since the public key is known to the attacker, this would allow the attacker to create a valid HMAC signature against this known key. To prevent this, the verifier needs to ensure that both the key material and the algorithm are appropriate for the usage in question. Additionally, while this specification does allow runtime specification of the algorithm using the `alg` signature parameter, applications are encouraged to use other mechanisms such as static configuration or higher protocol-level algorithm specification instead.
+
+## Parsing Structured Field Values {#security-structured}
+
+Several parts of this specification rely on the parsing of structured field values {{RFC8491}}. In particular, [normalization of HTTP structured field values](#http-header-structured), [referencing members of a dictionary structured field](#http-header-dictionary), and processing the `@signature-input` value when [verifying a signature](#verify). While structured field values are designed to be relatively simple to parse, a naive or broken implementation of such a parser could lead to subtle attack surfaces being exposed in the implementation.
+
+For example, if a buggy parser of the `@signature-input` value does not enforce proper closing of quotes around string values within the list of component identifiers, an attacker could take advantage of this and inject additional content into the signature input string through manipulating the `Signature-Input` field value on a message.
+
+To counteract this, implementations should use fully compliant and trusted parsers for all structured field processing, both on the signer and verifier side.
+
+## Choosing Message Components {#signature-components}
+
+Applications of HTTP Message Signatures need to decide which message components will be covered by the signature. Depending on the application, some components could be expected to be changed by intermediaries prior to the signature's verification. If these components are covered, such changes would, by design, break the signature. In such circumstances, it is often tempting for the verifier to simply turn off signature verification entirely in order to allow these otherwise good messages to pass. 
```

Comment:
This is the second time that you have mentioned "just turn it off".  I get that you want to acknowledge this tension that is created by security mechanisms that make things not work, but I think that you should be a little more careful in phrasing these to avoid creating the wrong impression.

Reply:
Ah thanks -- the other section was added after we wrote this paragraph, we can remove that text here. And I agree, it does need to be phrased carefully. However, I have seen too many security systems not acknowledge the fact that when security gets in the way of functionality, functionality naturally wins. I think that's particularly true for something like this, as described in the other section.